DemoApp/DemoAppCoordinator.swift

```diff
@@ -128,22 +132,44 @@ final class DemoAppCoordinator: NSObject, UNUserNotificationCenterDelegate {
         let controller: ChatChannelListController
         
         let connectCompletion: (Error?) -> Void = { [weak self] error in
+            guard let self = self else { return }
             if let error = error {
-                log.error("connecting the user failed \(error)")
-                if let self = self {
+                var errorTitle: String?
+                var errorMessage: String?
+                if let clientError = error as? ClientError,
+                   let underlyingError = clientError.underlyingError as? ClientError,
+                   let engineError = underlyingError.underlyingError as? WebSocketEngineError {
+                    if engineError.code == WebSocketEngineError.offlineErrorCode {
+                        log.info("connecting the user failed because we're offline")
+                    } else {
+                        errorTitle = "Connecting failed"
+                        errorMessage = engineError.reason
+                    }
+                } else if let clientError = error as? ClientError,
+                          let underlyingError = clientError.underlyingError as? ClientError,
+                          let errorPayload = underlyingError.underlyingError as? ErrorPayload {
```

Comment:
Would be nice to make this nested underlying error easier to access. Can't we have a `ClientError.errorPayload: ErrorPayload?` as well as a `ClientError.webSocketEngineError` ?

Reply:
I think at least we should create a ticket to try and improve the ClientError API 🤔

Reply:
this really needs a ticket since `ClientError.underlyingError` can be, and in this case _is_, another `ClientError`. so exposing `errorPayload` or anything would need ugly recursion